cart/application/cartService.go

```diff
@@ -945,13 +946,44 @@ func (cs *CartService) dispatchAllEvents(ctx context.Context, events []flamingo.
 
 // AdjustItemsToRestrictedQty checks the quantity restrictions for each item of the cart and returns what quantities have been adjusted
 func (cs *CartService) AdjustItemsToRestrictedQty(ctx context.Context, session *web.Session) (QtyAdjustmentResults, error) {
+	qtyAdjustmentResults, err := cs.GenerateRestrictedQtyAdjustments(ctx, session)
+	if err != nil {
+		return nil, err
+	}
+
 	cart, _, err := cs.cartReceiverService.GetCart(ctx, session)
 	if err != nil {
 		return nil, err
 	}
 
-	result := make(QtyAdjustmentResults, 0)
+	for _, qtyAdjustmentResult := range qtyAdjustmentResults {
+		couponCodes := cart.AppliedCouponCodes
+		if qtyAdjustmentResult.NewQty < 1 {
+			err = cs.DeleteItem(ctx, session, qtyAdjustmentResult.OriginalItem.ID, qtyAdjustmentResult.DeliveryCode)
+		} else {
+			err = cs.UpdateItemQty(ctx, session, qtyAdjustmentResult.OriginalItem.ID, qtyAdjustmentResult.DeliveryCode, qtyAdjustmentResult.NewQty)
+		}
+		if err != nil {
+			return nil, err
+		}
+		cart, _, err = cs.cartReceiverService.GetCart(ctx, session)
+		if err != nil {
+			return nil, err
+		}
+		qtyAdjustmentResult.HasRemovedCouponCodes = !couponCodes.ContainedIn(cart.AppliedCouponCodes)
+	}
 
+	return qtyAdjustmentResults, nil
+}
+
+// GenerateRestrictedQtyAdjustments checks the quantity restrictions for each item of the cart and returns which items should be adjusted and how
+func (cs *CartService) GenerateRestrictedQtyAdjustments(ctx context.Context, session *web.Session) (QtyAdjustmentResults, error) {
```

Comment:
Is this a useful exported/public function or only a helper which could be unexported?

Reply:
Unexported the helper function for now.